db/version_set.cc

```diff
@@ -5095,6 +5097,22 @@ Status VersionSet::ProcessManifestWrites(
           break;
         }
       }
+      // Version edit records above may contain WAL addition for WAL
+      // that already has deletion record in the manifest.
+      // Such addition record should be offsetted by a deletion record.
+      VersionEdit wal_deletions;
+      wal_deletions.DeleteWalsBefore(min_log_number_to_keep());
+      std::string wal_deletions_record;
+      if (!wal_deletions.EncodeTo(&wal_deletions_record)) {
+        s = Status::Corruption("Unable to Encode VersionEdit: " +
+                               wal_deletions.DebugString(true));
+      }
+      if (s.ok()) {
+        io_s = descriptor_log_->AddRecord(wal_deletions_record);
+        if (!io_s.ok()) {
+          s = io_s;
+        }
+      }
```

Comment:
A new `VersionEdit` on every MANIFEST update feels like a lot. Why do we need this again? I remember the `WriteCurrentStateToManifest()` was needed in case the addition is for a WAL that was deleted in the previous MANIFEST. Is this still needed for your test to pass after you added the logic to `WriteCurrentStateToManifest()`?

Reply:
Might be due to the UT failing without this - let me double check.

Reply:
Aha - okay here is the story:
Surprisingly the addition of obsoleted WAL didn't happen WriteCurrentStateToManifest() as the WAL addition in that function comes from VersionSet::wals_ https://github.com/facebook/rocksdb/blob/7.8.fb/db/version_set.cc#L5017, and wals_.AddWals() already has good sanitization in place like https://github.com/facebook/rocksdb/blob/7.8.fb/db/wal_edit.cc#L108.

The problematic addition comes from https://github.com/facebook/rocksdb/blob/7.8.fb/db/version_set.cc#L4931 and then https://github.com/facebook/rocksdb/blob/7.8.fb/db/version_set.cc#L5114. That was probably why I naively added the part you commented on.

But you are right - we don't need one VersionEdit per update - I fixed it now by check if such problematic wal addition exists first. 

And also I removed the special treatment in WriteCurrentStateToManifest() for reason mentioned in the beginning.

Will go thru make check and stress testing again before request for review again.

Reply:
I was aware the problematic AddWal() did not come from WriteCurrentStateToManifest(). But I thought the reason it was problematic was because the MANIFEST roll-over forgot to include a DeleteWalsBefore() entry. So I still have the same question about whether the WriteCurrentStateToManifest() change suffices to make the test pass.

Reply:
Ahh! I got it now. The prev UT did fail with only my WriteCurrentStateToManifest() change but it's because I mistakenly triggered it only if the problematic wals comes from VersionSet::wals_.

Now it's all clear to me and I will only have the WriteCurrentStateToManifest() treatment.

Reply:
Fixed - will go thru stress testing again.